src/models/tree/build_tree.cu

```diff
@@ -180,75 +181,356 @@ __global__ static void __launch_bounds__(BLOCK_THREADS)
+        ProcessPartialTileGlobal(offset, offset + kItemsPerTile);
+      } else {
+        ProcessTileShared(offset, tile_node_id);
+      }
+      offset += kItemsPerTile * gridDim.x;
+    }
+    ProcessPartialTileGlobal(offset, n_elements);
+
+    // Flush any remaining sums to the global histogram
+    shared_histogram.Flush(histogram, output);
+  }
+};
+
+template <typename T, int kBlockThreads, int kItemsPerThread>
+__global__ static void __launch_bounds__(kBlockThreads)
+  fill_histogram_shared(legate::AccessorRO<T, 3> X,
+                        int64_t sample_offset,
+                        legate::AccessorRO<double, 3> g,
+                        legate::AccessorRO<double, 3> h,
+                        size_t n_outputs,
+                        SparseSplitProposals<T> split_proposals,
+                        NodeBatch batch,
+                        Histogram<IntegerGPair> histogram,
+                        legate::Buffer<IntegerGPair, 2> node_sums,
+                        GradientQuantiser quantiser,
+                        legate::Buffer<int> feature_groups,
+                        int64_t seed)
+{
+  __shared__ SharedMemoryHistogramType shared_memory[kMaxSharedBins];
+  HistogramAgent<T, kBlockThreads, kItemsPerThread> agent(X,
+                                                          sample_offset,
+                                                          g,
+                                                          h,
+                                                          n_outputs,
+                                                          split_proposals,
+                                                          batch,
+                                                          histogram,
+                                                          node_sums,
+                                                          quantiser,
+                                                          feature_groups,
+                                                          seed,
+                                                          shared_memory);
+  agent.BuildHistogram();
+}
+
+// Manage the launch parameters for histogram kernel
+template <typename T, std::int32_t kBlockThreads = 1024, std::int32_t kItemsPerThread = 4>
+struct HistogramKernel {
+  const std::int32_t kItemsPerTile = kBlockThreads * kItemsPerThread;
+  legate::Buffer<int> feature_groups;
+  int num_groups;
+  int maximum_blocks_for_occupancy;
+  HistogramKernel(const SparseSplitProposals<T> split_proposals, cudaStream_t stream)
+  {
+    int device;
+    CHECK_CUDA(cudaGetDevice(&device));
+    std::int32_t n_mps = 0;
+    CHECK_CUDA(cudaDeviceGetAttribute(&n_mps, cudaDevAttrMultiProcessorCount, device));
+    std::int32_t n_blocks_per_mp = 0;
+    CHECK_CUDA(cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+      &n_blocks_per_mp,
+      fill_histogram_shared<T, kBlockThreads, kItemsPerThread>,
+      kBlockThreads,
+      0));
```

Comment:
Yes the occupancy is currently limited by registers. I definitely think it can be improved further. I actually don't observe any benefit from using more than 16kb shared memory, although performance gets worse if I use less.